src/dotnet/MLSearch.Service/Indexing/ChatIndexerWorker.cs

```diff
@@ -33,28 +73,27 @@ public async Task ExecuteAsync(int _shardIndex, CancellationToken cancellationTo
         // issues in case of re-sharding or new cluster rollouts.
         // It might have some other concurrent worker has updated
         // a cursor. TLDR: prevent stale cursor data.
-        while (!cancellationToken.IsCancellationRequested) {
-            try {
-                var e = await _channel
-                    .Reader
-                    .ReadAsync(cancellationToken)
-                    .ConfigureAwait(false);
-                var result = await dataIndexer.IndexNextAsync(e.Id, cancellationToken).ConfigureAwait(false);
-                if (!result.IsEndReached) {
-                    // Enqueue event to continue indexing.
-                    if (!Trigger.TryWrite(new MLSearch_TriggerChatIndexing(e.Id))) {
-                        Log.LogWarning("Event queue is full: We can't process till this indexing is fully complete.");
-                        while (!result.IsEndReached) {
-                            result = await dataIndexer.IndexNextAsync(e.Id, cancellationToken).ConfigureAwait(false);
-                        }
-                        Log.LogWarning("Event queue is full: exiting an element.");
-                    }
-                }
+
+        // TODO: add more logging and tracing
+        try {
```

Comment:
With this changes the indexing flow will not be even. It will first index a single chat and than and only then try to move to the next one. I've intentionally made it so it will try to keep up with all chats. 
This is open for a discussion though.

Reply:
This works in a separate logical thread, so other chats will be handled by other such threads in parallel.